spark/v3.2/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewriteDataFilesProcedure.java

```diff
@@ -0,0 +1,275 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.spark.extensions;
+
+import java.util.List;
+import java.util.Map;
+import org.apache.iceberg.AssertHelpers;
+import org.apache.iceberg.exceptions.ValidationException;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.spark.sql.AnalysisException;
+import org.apache.spark.sql.catalyst.analysis.NoSuchProcedureException;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Test;
+
+
+public class TestRewriteDataFilesProcedure extends SparkExtensionsTestBase {
+
+  public TestRewriteDataFilesProcedure(String catalogName, String implementation, Map<String, String> config) {
+    super(catalogName, implementation, config);
+  }
+
+  @After
+  public void removeTable() {
+    sql("DROP TABLE IF EXISTS %s", tableName);
+  }
+
+  @Test
+  public void testRewriteDataFilesInEmptyTable() {
+    sql("CREATE TABLE %s (id bigint NOT NULL, data string) USING iceberg", tableName);
+    List<Object[]> output = sql(
+        "CALL %s.system.rewrite_data_files('%s')", catalogName, tableIdent);
+    assertEquals("Procedure output must match",
+        ImmutableList.of(row(0, 0)),
+        output);
+  }
+
+  @Test
+  public void testRewriteDataFilesOnPartitionTable() {
+    sql("CREATE TABLE %s (id bigint NOT NULL, data string) USING iceberg PARTITIONED BY (data)", tableName);
+
+    // create 5 files for each partition (data = 'a' and data = 'b')
+    sql("INSERT INTO TABLE %s VALUES (1, 'a')", tableName);
```

Comment:
I would agree that we should just do a similar "createTable" sort of thing, in some internal tests I wrote I just copied the functions I added to rewrite

```
  private void createTable() {
    sql("DROP TABLE IF EXISTS %s", tableName);
    sql("CREATE TABLE %s (c1 int, c2 string, c3 string) using iceberg", tableName);
    List<ThreeColumnRecord> records1 = Lists.newArrayList();

    List<Integer> data = IntStream.range(0, 2000).boxed().collect(Collectors.toList());
    Collections.shuffle(data, new Random(42));
    data.forEach(i -> records1.add(new ThreeColumnRecord(i, "foo" + i, "bar" + i)));
    Dataset<Row> df = spark.createDataFrame(records1, ThreeColumnRecord.class).repartition(20);
    try {
      df.writeTo(tableName).append();
    } catch (NoSuchTableException e) {
      throw new RuntimeException(e);
    }
  }

  private void createPartitionedTable() {
    sql("DROP TABLE IF EXISTS %s", tableName);
    sql("CREATE TABLE %s (c1 int, c2 string, c3 string) using iceberg PARTITIONED BY (c1)", tableName);
    List<ThreeColumnRecord> records1 = Lists.newArrayList();

    List<Integer> data = IntStream.range(0, 2000).boxed().collect(Collectors.toList());
    Collections.shuffle(data, new Random(42));
    data.forEach(i -> records1.add(new ThreeColumnRecord(i % 5, "foo" + i, "bar" + i)));
    Dataset<Row> df = spark.createDataFrame(records1, ThreeColumnRecord.class).repartition(10);
    try {
      df.writeTo(tableName).append();
    } catch (NoSuchTableException e) {
      throw new RuntimeException(e);
    }
  }

  private void createBucketedTable() {
    sql("DROP TABLE IF EXISTS %s", tableName);
    sql("CREATE TABLE %s (c1 int, c2 string, c3 string) using iceberg PARTITIONED BY (bucket(4, c3))", tableName);
    List<ThreeColumnRecord> records1 = Lists.newArrayList();

    List<Integer> data = IntStream.range(0, 2000).boxed().collect(Collectors.toList());
    Collections.shuffle(data, new Random(42));
    data.forEach(i -> records1.add(new ThreeColumnRecord(i % 5, "foo" + i, "bar" + i)));
    Dataset<Row> df = spark.createDataFrame(records1, ThreeColumnRecord.class).repartition(10);
    try {
      df.writeTo(tableName).append();
    } catch (NoSuchTableException e) {
      throw new RuntimeException(e);
    }
  }```
  
  Although I don't think bucketed would work here

Reply:
we should have a common util for this ? Instead of duplicating this code in test classes ?

Reply:
with the given sample here, create partition table fails with `Encountered records that belong to already closed files:` error due to `i%5`
If I remove modulo operator each partition will have only one file. It is getting complicated for my scenario. I will just optimize by current test case by extracting lines and putting it in a loop as @rymurr suggested.

Reply:
@RussellSpitzer : I was pointing links to this comment about the problems I faced while using the dataframe

Reply:
Just set it to fan out write enabled, and you won't have that error.

Reply:
done..